### PR TITLE
ref(starfish): More sophisticated API typing

### PR DESCRIPTION
**Pull request by gggritso**
tl;dr this PR introduces a new set of types that makes it possible to have much more sophisticated typing for API responses and improves DX.

## Changes

I wanted to accomplish two goals:

- it should be impossible to typo a field name (e.g., asking to fetch `spss()` should error)
- the server response types should take the fields into account (e.g., if I asked for `sps()`, the response should be typed as having that field, and no others)

To do this, I introduced a new typing setup, and applied it to `getSpanMetrics` as an example. The wins are immediate!

**e.g.,**

Autocomplete for requested fields is helpful and exhaustive.
Screenshot 2023-08-30 at 5 44 44 PM

It is illegal to try to fetch a field that doesn't exist.

Screenshot 2023-08-30 at 5 45 02 PM

The server response is _typed_. Only the fields that were requested are available on the returned object! You can't get `data['sps()']` if you didn't request that field.

Screenshot 2023-08-30 at 5 45 20 PM

The server response fields have correct types (this was already the case, but it's cleaner now).

Screenshot 2023-08-30 at 5 45 56 PM

## Next Steps

This also has really good implications for the code!

1. No need to use enums like `SpanFunction.SPS` anymore. It's _much_ more ergonomic to use plain strings in most cases. If TypeScript's autocomplete is autofilling `sps()`, it sucks to have to import some other constant to do it instead of just using the autocomplete! I chose not to remove enums yet, but I think removing them is a pretty good idea. Fewer imports, much more readable code.
2. We can remove a lot of our manual types, since types are inferred correctly from server responses. That, or we can use `Pick` on `MetricsResponse`

IMO this is a pretty good first step. We can apply the same approach to our other hooks and eliminate a whole class of runtime errors that we ran into when we first released this internally.

### CODE REVIEW

File: static/app/views/starfish/queries/useSpanMetrics.tsx [hunk 1]
```diff
@@ -3,29 +3,24 @@ import {Location} from 'history';
...
-  'time_spent_percentage()': number;
-};
-
 export type SpanSummaryQueryFilters = {
   'transaction.method'?: string;
   transactionName?: string;
 };
 
-export const useSpanMetrics = (
+export const useSpanMetrics = <T extends MetricsProperty[]>(
```
Comment (gggritso) [inline static/app/views/starfish/queries/useSpanMetrics.tsx:20]: `<T extends MetricsProperty[]>` is a type constraint. This ensures that the `fields` argument is a list of metrics properties.

File: static/app/views/starfish/queries/useSpanMetrics.tsx [hunk 2]
```diff
@@ -37,14 +32,20 @@ export const useSpanMetrics = (
...
-  const result = useSpansQuery<SpanMetrics[]>({
+  const result = useSpansQuery({
     eventView,
     initialData: [],
     enabled,
     referrer,
   });
 
-  return {...result, data: result?.data?.[0] ?? {}, isEnabled: enabled};
+  const data = (result?.data?.[0] ?? {}) as Pick<MetricsResponse, T[number]>;
```
Comment (gggritso) [inline static/app/views/starfish/queries/useSpanMetrics.tsx:42]: I didn't have a chance to update the upstream typing, so I did `as` here to coerce it. It would be even better to type the whole hook so the code inside has to conform to the contract it declares

Comment (gggritso) [reply static/app/views/starfish/queries/useSpanMetrics.tsx:42]: `T[number]` sets the type of the returned data to the type of an elements inside `T`, which is the exact value of the `fields` argument

File: static/app/views/starfish/types.tsx [hunk 3]
```diff
@@ -28,17 +28,38 @@ export enum SpanMetricsField {
...
+export type MetricsResponse = {
+  [Property in SpanNumberFields as `avg(${Property})`]: number;
+} & {
+  [Property in SpanNumberFields as `sum(${Property})`]: number;
+} & {
+  [Property in SpanFunctions as `${Property}()`]: number;
+} & {
+  [Property in SpanStringFields as `${Property}`]: string;
+} & {
+  'time_spent_percentage(local)': number;
```
Comment (gggritso) [inline static/app/views/starfish/types.tsx:58]: These are [template literal types](https://www.typescriptlang.org/docs/handbook/2/template-literal-types.html)

File: static/app/views/starfish/views/spanSummaryPage/index.tsx [hunk 4]
```diff
@@ -64,24 +63,16 @@ function SpanSummaryPage({params, location}: Props) {
   const {data: spanMetrics, isLoading: isSpanMetricsLoading} = useSpanMetrics(
     groupId,
     queryFilter,
-    [
-      SpanMetricsField.SPAN_OP,
-      SpanMetricsField.SPAN_GROUP,
-      SpanMetricsField.PROJECT_ID,
-      `${SpanFunction.SPS}()`,
-    ],
+    ['span.op', 'span.group', 'project.id', 'sps()'],
```
Comment (gggritso) [inline static/app/views/starfish/views/spanSummaryPage/index.tsx:66]: Using plain strings is more readable and dense in a good way. Since the argument is typed, it's not possible to typo, so the enum is not really necessary.

File: static/app/views/starfish/views/spanSummaryPage/spanTransactionsTable.tsx [hunk 5]
```diff
@@ -50,10 +50,7 @@ type Row = {
 
 type Props = {
   sort: ValidSort;
-  span: Pick<
-    SpanMetricsFieldTypes,
-    SpanMetricsField.SPAN_GROUP | SpanMetricsField.SPAN_OP
-  >;
+  span: Pick<MetricsResponse, SpanMetricsField.SPAN_GROUP | SpanMetricsField.SPAN_OP>;
```
Comment (gggritso) [inline static/app/views/starfish/views/spanSummaryPage/spanTransactionsTable.tsx:53]: `MetricsResponse` is probably a more descriptive way to describe the type anyway

File: static/app/views/starfish/views/spanSummaryPage/index.tsx (hunk 4 above)
Comment (AbhiPrasad) [reply static/app/views/starfish/views/spanSummaryPage/index.tsx:66]: https://blog.sentry.io/performance-impact-of-generated-javascript/#switching-from-typescript-enums-to-const-and-string-enums 👀

**Review by AbhiPrasad — APPROVED** [review]
thanks George!
